src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -273,6 +422,9 @@ object InModuleBody
         result.get
       }
     }
+
+    // Store [[out.execute]] to [[scope.inModuleBody]],
+    // it is a val with type of `() => Unit`, which will be executed in [[LazyModuleImp.instantiate]].
```

Comment:
I'm still curious what the order of evaluation is here. It seems like things that are added later actually get evaluated first? So it's not last-connect semantics? Is that correct @terpstra ?

Reply:
As far as I can tell, the items in `inModuleBody` are always evaluated in the order that they are pushed onto the stack. The only other instance of `inModuleBody` I see is this line, where the list is reversed before we iterate over them: https://github.com/chipsalliance/rocket-chip/blob/a52fbb2eb077fd42fdf23a6f4e4c7d20dea00142/src/main/scala/diplomacy/LazyModule.scala#L176

This is likely done for performance reasons, since it's O(1) for pushing (prepending) to the list and it's O(n) for iterating over the final list, which will happen exactly once in any execution of Diplomacy.

Reply:
right... but it changes the functionality I think? users need to be aware of the difference in behavior if you have several `InModuleBody` calls in a module (e.g. if you have lots of inheritance).

Reply:
What functionality are you expecting? My reading of this code is that all the `InModuleBody` things get evaluated in the same order that they are pushed onto this stack, so it's FIFO order. Is that not the intuitive behavior?

Reply:
somewhere there is a `reverse` in this code which I don't understand why it is there. Are you looking at that and that is why...?

Reply:
https://github.com/chipsalliance/rocket-chip/pull/2311/files#diff-48513a3a870e4bcb7999b3e75261d9c9R308

Reply:
oh you already mentioned that line above.